presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java

```diff
@@ -103,19 +105,28 @@ public long getInMemorySizeInBytes()
     }
 
     @Override
-    public long getJoinPosition(int position, Page page)
+    public long getJoinPosition(int position, Page hashChannelsPage, Page allChannelsPage)
     {
-        return getJoinPosition(position, page, pagesHashStrategy.hashRow(position, page.getBlocks()));
+        return getJoinPosition(position, hashChannelsPage, allChannelsPage, pagesHashStrategy.hashRow(position, hashChannelsPage.getBlocks()));
     }
 
     @Override
-    public long getJoinPosition(int position, Page page, int rawHash)
+    public long getJoinPosition(int rightPosition, Page hashChannelsPage, Page allChannelsPage, int rawHash)
     {
         int pos = getHashPosition(rawHash, mask);
 
         while (key[pos] != -1) {
-            if (positionEqualsCurrentRow(key[pos], position, page.getBlocks())) {
-                return key[pos];
+            if (positionEqualsCurrentRow(key[pos], rightPosition, hashChannelsPage.getBlocks(), allChannelsPage.getBlocks())) {
+                long candidate = key[pos];
+                if (filterFunctionPresent) {
+                    while (candidate != -1 && !checkFilterFunctionCurrentRow(Ints.checkedCast(candidate), rightPosition, allChannelsPage.getBlocks())) {
+                        candidate = getNextJoinPosition(candidate, rightPosition, allChannelsPage);
```

Comment:
extract common code to something like `getFirstValidPositionAfter(...)` as we talked.